src/csv_processor.cpp

```diff
@@ -1,25 +1,268 @@
+  }
+
+  constexpr int MAGIC_SIZE = 4;
+  uint8_t magic_bytes[MAGIC_SIZE];
+  ifs.read(reinterpret_cast<char *>(magic_bytes), sizeof(magic_bytes));
+
+  if (ifs.fail() && !ifs.eof()) {
+    throw std::system_error(
+        errno, std::generic_category(),
+        "Failed trying to read zstd magic bytes from file <" + file_path + ">");
+  }
+
+  // File has fewer than 4 bytes, hence cannot be zstd-compressed
+  if (ifs.gcount() < MAGIC_SIZE) {
+    return CompressionType::None;
+  }
+
+  // Check for ZSTD magic number (0x28B52FFD in little-endian)
+  const bool is_zstd_compressed =
+      magic_bytes[0] == 0x28 && magic_bytes[1] == 0xB5 &&
+      magic_bytes[2] == 0x2F && magic_bytes[3] == 0xFD;
+  return is_zstd_compressed ? CompressionType::ZSTD : CompressionType::None;
+}
+
+/// Adds a SELECT clause with the specified columns to the query
+void add_projections(std::ostringstream &query,
+                     const std::vector<column_def> &columns) {
+  query << " SELECT";
+
+  if (columns.empty()) {
+    query << " *";
+  } else {
+    for (const auto &column : columns) {
+      // DuckDB can handle trailing commas
+      query << " " << duckdb::KeywordHelper::WriteQuoted(column.name, '"')
+            << ",";
+    }
+  }
+}
+
+/// Adds CSV reader options related to column types to the query (all_varchar or
+/// column_types)
+void add_type_options(std::ostringstream &query,
+                      const std::vector<column_def> &columns,
+                      const bool allow_unmodified_string) {
+  // We set all_varchar=true if we have to deal with `unmodified_string`. Those
+  // are string values that represent an unchanged value in an UPDATE or UPSERT,
+  // and they break type conversion in the CSV reader. DuckDB does an implicit
+  // conversion later during the UPDATE/UPSERT. If `unmodified_string` is not
+  // set (for UPSERT and DELETE), we push down the type conversion to the CSV
+  // reader.
+  if (allow_unmodified_string) {
+    query << ", all_varchar=true";
     return;
   }
-  throw std::invalid_argument("File <" + file_path +
-                              "> is missing or inaccessible");
+
+  bool has_valid_column_types = false;
+  for (const auto &column : columns) {
+    if (column.type != duckdb::LogicalTypeId::INVALID) {
+      has_valid_column_types = true;
+      break;
+    }
+  }
+
+  if (!has_valid_column_types) {
+    // No valid column types. We need to back out because column_types must not
+    // be an empty struct.
+    return;
+  }
+
+  // We cannot assume the order of columns. From the Fivetran Partner SDK docs:
+  // "Always read the CSV file header to determine the column order."
+  // Therefore, we do not set the `columns` parameter, but `column_types` which
+  // is a mapping from a column name to a type:
+  // column_types={'colB':'VARCHAR','colA':'INTEGER'}.
+  // DuckDB detects the order of columns by reading the header row.
+  // If no columns are specified, DuckDB will auto-detect all column types.
```

Comment:
I guess that answers my question? So the columns property should be in the order of the columns in the target, and this part takes care of the mapping?